akka-http-core/src/main/scala/akka/http/scaladsl/model/ErrorInfo.scala

```diff
@@ -5,22 +5,26 @@
 package akka.http.scaladsl.model
 
 import StatusCodes.ClientError
+import akka.annotation.InternalApi
 
 /**
+ * INTERNAL API
  * Two-level model of error information.
  * The summary should explain what is wrong with the request or response *without* directly
  * repeating anything present in the message itself (in order to not open holes for XSS attacks),
  * while the detail can contain additional information from any source (even the request itself).
  */
-final case class ErrorInfo(summary: String = "", detail: String = "") {
+@InternalApi private[akka] final case class ErrorInfo(summary: String = "", detail: String = "", errorHeaderName: String = "") {
```

Comment:
We cannot simply make it an internal class as far as I can see, it is part of the public API in `IllegalRequestException` for example.

Keeping this binary compatible while adding a field requires quite a bit of gymnastics I'm afraid, if completely possible at all: we need a two param constructor, a two param `apply`, etc.

Reply:
Sorry to give you conflicting feedback btw, I see you got directed to annotate like this, maybe @ktoso know something I don't know.

Reply:
Thank you for review.

I understand the reason for can't change internal class.
I missed it is part of the public API. And thank you for pointing that out.

How about the following method to keep case class and binary compatible? Although a bit tricky.

```scala
final case class ErrorInfo(summary: String = "", detail: String = "") {
  var errorHeaderName = ""
  def withSummary(newSummary: String) = copy(summary = newSummary).withErrorHeaderName(errorHeaderName)
  def withSummaryPrepended(prefix: String) = withSummary(if (summary.isEmpty) prefix else prefix + ": " + summary)
  def withErrorHeaderName(headerName: String) = {
    errorHeaderName = headerName.toLowerCase
    this
  }
  def withFallbackSummary(fallbackSummary: String) = if (summary.isEmpty) withSummary(fallbackSummary) else this
  def formatPretty = if (summary.isEmpty) detail else if (detail.isEmpty) summary else summary + ": " + detail
  def format(withDetail: Boolean): String = if (withDetail) formatPretty else summary
}
```

Or should we rethink from the idea of adding headerName to ErrorInfo.
WDYT @johanandren @ktoso ?

Reply:
I don't think we can have a `var` in there because of concurrency. Otoh, maybe we can make the constructor internal, and keep the rest of the binary signature intact somehow? The type and accessors needs to be public but there is not really a reason for anyone outside of the library to construct instances.

Reply:
I indeed did not consider concurrency.
Thank you for your explanation.
I understood the reason why I should not use `var`. 

And I'm sorry. I missed your proposal that you wrote above.

Does the meaning that you wrote match in the following?

1. add two param `apply` method and mark internal. As described below.

```scala
object ErrorInfo {
  /**
   * Allows constructing an `ErrorInfo` from a single string.
   * Used for example when catching exceptions generated by the header value parser, which doesn't provide
   * summary/details information but structures its exception messages accordingly.
   */
  def fromCompoundString(message: String): ErrorInfo = message.split(": ", 2) match {
    case Array(summary, detail) ⇒ apply(summary, detail)
    case _                      ⇒ ErrorInfo("", message)
  }

  /** INTERNAL API */
  @InternalApi private[akka] def apply(summary: String, detail: String): ErrorInfo = {
    ErrorInfo(summary, detail)
  }
}
```

2. `copy` and `this` methods are still binary incompatible. I still don't understand well this...
Should we add two methods to mima-filters configuration? Or add methods just like `apply`?

Reply:
`this`/constructor can be solved with auxiliary constructor, but copy method is impossible to get around I think, since it is generated for case classes. Maybe there is no way around breaking bin comp. I think that if users touch this it will be the unapply, could we keep a two param unapply somehow so that wouldn't be incompatible at least?

Reply:
Thank you for replay.

I understand that how to fix bin comp for `this method` and `copy method`.
And indeed it seems better to consider about `unapply method`, too.

I defined unapply method as below.

```scala
object ErrorInfo {
  /** INTERNAL API */
  @InternalApi private[akka] def apply(summary: String, detail: String): ErrorInfo = {
    ErrorInfo(summary, detail, "")
  }

  def unapply(info: ErrorInfo): Option[(String, String)] = Some((info.summary, info.detail))
  ...
```

But it did not work well.

```scala
scala> val info = ErrorInfo("summary", "detail")

info: akka.http.scaladsl.model.ErrorInfo = ErrorInfo(summary,detail,)

scala> val ErrorInfo(summary, detail) = info
<console>:18: error: wrong number of arguments for pattern akka.http.scaladsl.model.ErrorInfo(summary: String,detail: String,errorHeaderName: String)
       val ErrorInfo(summary, detail) = info
```

`unapply method` generated By case class and `unapply method` I defined both have the same arguments, although the return value is different. 
Therefore, I think that it is probably impossible to decide which is used.

For now, I pushed fixed code without unapply.